File: cpp/147_Insertion_Sort_List.cpp
/*
Sort a linked list using insertion sort.
A graphical example of insertion sort. The partial sorted list (black) initially contains only the first element in the list.
With each iteration one element (red) is removed from the input data and inserted in-place into the sorted list
 

Algorithm of Insertion Sort:

Insertion sort iterates, consuming one input element each repetition, and growing a sorted output list.
At each iteration, insertion sort removes one element from the input data, 
finds the location it belongs within the sorted list, and inserts it there.
It repeats until no input elements remain.

Example 1:
Input: 4->2->1->3
Output: 1->2->3->4

Example 2:
Input: -1->5->3->4->0
Output: -1->0->3->4->5
*/
#include <iostream>

using namespace std;

class Solution {
    public:
        struct ListNode {
            int val;
            ListNode *next;
            ListNode(int x) : val(x), next(NULL) {}
        };

        ListNode* insertionSortList(ListNode* head) {
            if(!head || !head->next) return head;
            ListNode *dummy = new ListNode(0);
            dummy->next = head;
            ListNode *pre = head, 
                     *cur = head->next, 
                     *h = head, 
                     *next;
            while(cur) {
                if(cur->val < pre->val) {
                    next = cur->next;
                    pre->next = next;
                    ListNode *p = nullptr;
                    while(cur->val > h->val) {
                        p = h;
                        h = h->next;
                    }
                    if(p) {
                        p->next = cur;
                    } else {
                        dummy->next = cur;
                    }
                    cur->next = h;
                    h = dummy->next;
                    cur = next;
                } else {
                    pre = cur;
                    cur = cur->next;
                }
            }
            return dummy->next;
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/147_Insertion_Sort_List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        dummy = ListNode(0)
        dummy.next, node, prev = head, head.next, head
        while node:
            if prev.val > node.val:
                next = node.next
                prev.next = next
                tmp = None
                while node.val > head.val:
                    tmp = head
                    head = head.next
                if tmp:
                    tmp.next = node
                else:
                    dummy.next = node
                node.next = head
                head = dummy.next
                node = next
            else:
                prev = node
                node = node.next
        return dummy.next
